server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -240,19 +246,51 @@ public AppenderatorAddResult add(
     final int numAddedRows = sinkRowsInMemoryAfterAdd - sinkRowsInMemoryBeforeAdd;
     rowsCurrentlyInMemory.addAndGet(numAddedRows);
     totalRows.addAndGet(numAddedRows);
+    bytesCurrentlyInMemory.addAndGet(bytesInMemoryAfterAdd - bytesInMemoryBeforeAdd);
 
     boolean isPersistRequired = false;
-    if (!sink.canAppendRow()
-        || System.currentTimeMillis() > nextFlush
-        || rowsCurrentlyInMemory.get() >= tuningConfig.getMaxRowsInMemory()) {
+    boolean persist = false;
+    List<String> persistReasons = Lists.newArrayList();
+
+    if (!sink.canAppendRow()) {
+      persist = true;
+      persistReasons.add("No more rows can be appended to sink");
+    }
+    if (System.currentTimeMillis() > nextFlush) {
+      persist = true;
+      persistReasons.add(StringUtils.format(
+          " current time[%d] is greater than nextFlush[%d],",
+          System.currentTimeMillis(),
+          nextFlush
+      ));
+    }
+    if (rowsCurrentlyInMemory.get() >= tuningConfig.getMaxRowsInMemory()) {
+      persist = true;
+      persistReasons.add(StringUtils.format(
+          " rowsCurrentlyInMemory[%d] is greater than maxRowsInMemory[%d],",
+          rowsCurrentlyInMemory.get(),
+          tuningConfig.getMaxRowsInMemory()
+      ));
+    }
+    if (tuningConfig.getMaxBytesInMemory() != -1
```

Comment:
Better to be `tuningConfig.getMaxBytesInMemory() > 0`

Reply:
done